File: netDxf/IO/DxfWriter.LayerFilterPointer.cs
using System.Linq;
using netDxf.Collections;
using netDxf.Objects;

namespace netDxf.IO
{
    internal sealed partial class DxfWriter
    {
        private void PrepareLayerFilterPointerClasses(DxfClassCollection definitions)
        {
            this.PrepareStoredEnvelopeClass(definitions, "LAYER_FILTER", "AcDbLayerFilter", "ObjectDBX Classes", 0,
                this.doc.Objects.Items.Any(item => item is DxfLayerFilter));
            // Autodesk supplies the C++ name and flags, but no application name for OBJECT_PTR.
            this.PrepareStoredEnvelopeClass(definitions, "OBJECT_PTR", "CAseDLPNTableRecord", string.Empty, 1,
                this.doc.Objects.Items.Any(item => item is DxfObjectPointer));
        }

        private void PrepareStoredEnvelopeClass(DxfClassCollection definitions, string name, string cppName,
            string applicationName, int flags, bool typedPresent)
        {
            int count = this.doc.Objects.Items.Count(item => item.CodeName == name);
            if (!typedPresent)
            {
                // A compatible declaration can outlive its last typed instance after erasure.
                // Unfamiliar declarations retain their private metadata unchanged.
                if (definitions.Contains(name) && definitions[name].CppClassName == cppName && !definitions[name].IsEntity)
                    definitions[name].InstanceCount = count;
                return;
            }
            if (definitions.Contains(name))
            {
                DxfClass definition = definitions[name];
                if (definition.CppClassName != cppName || definition.IsEntity)
                    throw new System.IO.InvalidDataException("CLASS conflicts with a typed database object: " + name);
                definition.InstanceCount = count;
            }
            else definitions.Add(new DxfClass(name, cppName, applicationName)
                { ProxyFlags = flags, IsEntity = false, WasProxy = false, InstanceCount = count });
        }

        private bool WriteLayerFilterPointerPayload(DxfDatabaseObject item)
        {
            if (item is DxfLayerFilter filter)
            {
                this.chunk.Write(100, "AcDbFilter");
                this.chunk.Write(100, "AcDbLayerFilter");
                foreach (string name in filter.LayerNames) this.chunk.Write(8, this.EncodeDatabaseString(name));
                return true;
            }
            return item is DxfObjectPointer; // No public subclass or pointer payload.
        }
    }
}

File: javascript/netDxf/IO/DxfWriter.LayerFilterPointer.js
// Copyright (c) netDxf contributors. MIT License; see package LICENSE.
import { DxfClass } from '../DxfClass.js';
import { DxfLayerFilter } from '../Objects/DxfLayerFilter.js';
import { DxfObjectPointer } from '../Objects/DxfObjectPointer.js';
import { EncodeDxfDatabaseText } from '../../runtime/DxfStringEncoding.js';
import { InvalidDataException } from '../../runtime/Errors.js';
export function PrepareStoredEnvelopeClass(document,definitions,name,cppName,applicationName,flags,typedPresent){
  const count=Array.from(document.Objects.Items).filter(item=>item.CodeName===name).length;
  if(!typedPresent){
    if(definitions.Contains(name)&&definitions.get_Item(name).CppClassName===cppName&&!definitions.get_Item(name).IsEntity)definitions.get_Item(name).InstanceCount=count;return;
  }
  if(definitions.Contains(name)){
    const definition=definitions.get_Item(name);if(definition.CppClassName!==cppName||definition.IsEntity)throw new InvalidDataException('CLASS conflicts with a typed database object: '+name);definition.InstanceCount=count;
  }else{const definition=new DxfClass(name,cppName,applicationName);definition.ProxyFlags=flags;definition.IsEntity=false;definition.WasProxy=false;definition.InstanceCount=count;definitions.Add(definition);}
}
export function PrepareLayerFilterPointerClasses(document,definitions){
  PrepareStoredEnvelopeClass(document,definitions,'LAYER_FILTER','AcDbLayerFilter','ObjectDBX Classes',0,Array.from(document.Objects.Items).some(item=>item instanceof DxfLayerFilter));
  PrepareStoredEnvelopeClass(document,definitions,'OBJECT_PTR','CAseDLPNTableRecord','',1,Array.from(document.Objects.Items).some(item=>item instanceof DxfObjectPointer));
}
export function WriteLayerFilterPointerPayload(chunk,version,item){
  if(item instanceof DxfLayerFilter){chunk.Write(100,'AcDbFilter');chunk.Write(100,'AcDbLayerFilter');for(const name of item.LayerNames)chunk.Write(8,EncodeDxfDatabaseText(name,version));return true;}
  return item instanceof DxfObjectPointer;
}
